### PR TITLE
Supplementary reads should contribute to metrics that count bases in

### PR DESCRIPTION
**Pull request by nh13**
`AlignmentSummaryMetrics`, `RnaSeqMetrics`, and `QualityYieldMetrics`.

Note that the same update for `HsMetrics`, `TargetedMetrics`, and `TargetedPcrMetrics` is included in the following pull request: https://github.com/broadinstitute/picard/pull/332

### CODE REVIEW

**Comment by nh13** [comment]
@tfenne @yfarjoun I would appreciate a review on this one.  The goal was to use supplementary alignments for metrics that count the number of bases, but not for metrics that count the number of reads.

**Comment by tfenne** [comment]
I took a quick look.  My initial reaction is that the change to CollectQualityYieldMetrics is incorrect.  Going one further I think the goal of this PR should be restated as "to use supplementary alignments for metrics that count the number of *aligned* bases, but not for metrics that count the number of reads".

To some degree this is tough because it depends on how the aligner and any other processing has treated records with split alignments.  I've only deal with bwa's interpretation, and there I've only seen it handle thing by 
a) Leaving one (non-supplemental) record with all the bases and quality scores, but soft-clipping the alignment
b) Hard or soft-clipping the supplemental alignments

So when counting aligned bases you should sum across the two, but when counting just _bases_ you'd want to exclude the supplementals again.

**Comment by yfarjoun** [comment]
your last "bases" should read "reads", yes?

**Comment by tfenne** [comment]
@yfarjoun - no.  That's the point I was trying to make.  I think it is:
counting "reads": exclude supplementals
counting "aligned bases" (and derived): include supplementals
counting "unaligned bases" (and derived): exclude supplementals

**Comment by nh13** [comment]
@tfenne @yfarjoun fixed some tests and your issue in `CollectQualityYieldMetrics`

**Comment by yfarjoun** [comment]
Ah, I see.  And, yes, I agree.

**Comment by nh13** [comment]
@yfarjoun I don't have permission to merge, would you do the honors?

File: src/java/picard/analysis/directed/RnaSeqMetricsCollector.java [hunk 1]
```diff
@@ -113,17 +113,23 @@ public PerUnitRnaSeqMetricsCollector(final String sample,
...
+            // NB: for read count metrics, do not include supplementary records, but for base count metrics, do include supplementary records.
+
             // Filter out some reads, and collect the total number of PF bases
-            if (rec.getReadFailsVendorQualityCheckFlag() || rec.isSecondaryOrSupplementary()) return;
+            if (rec.getReadFailsVendorQualityCheckFlag()) return;
 
-            this.metrics.PF_BASES += rec.getReadLength();
-            if (rec.getReadUnmappedFlag()) return;
+            // NB: we count unmapped reads here
+            if (!rec.getNotPrimaryAlignmentFlag()) this.metrics.PF_BASES += rec.getReadLength();
```
Comment (yfarjoun) [inline src/java/picard/analysis/directed/RnaSeqMetricsCollector.java]: You're allowing secondary alignments here?

Comment (yfarjoun) [inline src/java/picard/analysis/directed/RnaSeqMetricsCollector.java]: nm, found it.

**Comment by ktibbett** [comment]
@nh13 @tfenne I need to get a better understanding of how this will impact production pipelines; working on this now.

**Comment by tfenne** [comment]
Thanks @ktibbett .  Let us know if we can do anything to help.

**Comment by ktibbett** [comment]
Sorry for the delay.  :+1:

**Comment by tfenne** [comment]
Thanks @ktibbett 
@nh13 Can you please rebase on top of master and re-push?  Once tests pass again I'll merge.

**Comment by nh13** [comment]
@tfenne rebased; it's ready to merge when the tests pass.

**Pull request merged**